Implementation of DiscordMessage.Content humanization (attempt number 2)

Evengard opened the pull request:
# Summary
This is the implementation of the convertation to a human-readable format of DiscordMessage snowflake tags (such as mentions and emojis). Resolves #935. This is a second (and probably final) attempt, in a hope that this way of implementing stuff would be more appropriate for the project.

# Details
For some cases, a human-readable version of the message received by a bot may be useful - for example, to relay it somewhere else, eg on a website. This pull request implements the convertation to a human-readable format the most cryptic part of a bot-received message - the mentions and emojis.
Now, the way I did the first PR were probably not right, sorry - I didn't knew that. I genuinely try to do it right, I'm just new to this project.

# Changes proposed
* The implementation changed completely from extending Discord objects to a self contained Utility function doing everything inside itself. Only small changes are made to DiscordEmoji.EmojiUtils.cs by making the arrays internal instead of private.
* Added the `Utilities.HumanizeContentAsync` method for converting the snowflake tags into human-readable representation
* Because the method of converting emojis may depend on circumstances, `HumanizeContentAsync` accepts a second parameter of `Func<DiscordEmoji, Task<string>>` to allow specifying a method of formatting theese
* The default is to skip all non-unicode emojis, and output the unicode emojis in unicode format
* Sometimes the guild emoji used in a message can't be retrieved using default Discord API because the bot doesn't know about the server from which the emote originate (eg a Nitro user posted an emoji from another server). We still have enough information to construct a dummy DiscordEmoji object which could be used for retrieving the URL of the emote, for example. We explicitely mark it as IsAvailable = false to avoid it's usage elsewhere
* Along the way fixed the Formatter.Strip to account for animated emojis as well

# Notes
The `HumanizeContentAsync` method is async only because of the `message.Channel.Guild.GetMemberAsync` call, which is in turn a fallback for the case if `if (user is DiscordMember member)` fails. Now, if for guild messages this cast is always true, I can rework it to make it non-async.

Evengard commented:
Now, if there is a need to avoid using DiscordMessage.MentionedUsers/MentionedRoles/MentionedChannels because of planned architectural changes, I can rework that too. I am trying to be open and flexible about this PR, so feel free to ping me. For possible discussion I joined the project's Discord server (Evengard#2406), so feel free to ping me here too.

Evengard commented:
I've ended up reworking it, making it more self-contained than ever.
* Not using DiscordMessage.MentionedUsers/MentionedRoles/MentionedChannels because of possible architectural changes to the library
* Transformed the callback to be more flexible - it now accepts not only DiscordEmoji, but SnowflakeObject, which allows to reuse it for formatting other mentions, not only emojis
* The default implementation outputs @DisplayName for nickname mentions, @UserName for user mentions, @RoleName for role mentions, #ChannelName for channel mentions

Note: The method now uses more async methods, so seems that reworking to a non-async method is not an option. I can change the callback to be non-async though.

VelvetThePanda commented on DSharpPlus/Utilities.cs:
```diff
@@ -169,12 +169,154 @@ internal static IEnumerable<ulong> GetChannelMentions(DiscordMessage message)
+            // Now we take care of common emojis
+            // We sort them so we first replace the most complex ("composite") ones first
+
+            // Converting the ":emojiname:" format
+            var emojis = DiscordEmoji.UnicodeEmojis.Keys
+                .Where(e => text.Contains(e))
+                .OrderByDescending(e => e.Count(c => c == ':'))
+                .ThenByDescending(e => e.Length)
+                .Select(e => new { Token = e, Emoji = DiscordEmoji.FromName(message.Discord, e, false) })
+                .Union(// Converting the unicode emojis
+                    DiscordEmoji.DiscordNameLookup.Keys
+                    .Where(e => text.Contains(e))
+                    .OrderByDescending(e => e.Length)
+                    .Select(e => new { Token = e, Emoji = DiscordEmoji.FromUnicode(e) }));
+
+            foreach (var emojiInfo in emojis)
+            {
+                var replacement = await objectReplacer(emojiInfo.Emoji);
+                text = text.Replace(emojiInfo.Token, replacement);
+            }
+
+            return text;
```
This is an abomination. Simply escaping both colons and stripping the Id (if applicable) would be infinitely faster and more readable.

Evengard replied:
The ID here is not applicable, the Ided guild emotes were parsed already at this point, this is for common ones.
I wanted to make sure I catch all the emojis so they can be all passed down to the callback for user-defined replacement, and not catching something which shouldn't (eg :somerandomtext: not being an emote).

VelvetThePanda replied:
If you use Regex.Replace it doesn't matter if it's a valid emote or not because the point isn't to check if an emote is valid, if you're just trying to humanize what could be an emote

Evengard replied:
It's a little complex than that. I need to make sure that the output is as close as possible to what is shown by Discord itself to humans. So if I replace an invalid emote with something which isn't replaced by Discord, it wouldn't be correct.

VelvetThePanda replied:
Discord renders the emoji, so. And escaping an invalid emoji produces the same result as escaping a valid emoji.

Evengard replied:
Well, I ditched all that code. ¯\\\_(ツ)\_/¯

VelvetThePanda commented on DSharpPlus/Utilities.cs:
```diff
@@ -289,6 +431,33 @@ public static bool Contains(this string str, params char[] characters)
+        /// <returns>A new string that is identical to the input string, except that a replacement
+        /// string takes the place of each matched string. If pattern is not matched in the
+        /// current instance, the method returns the current instance unchanged</returns>
+        public static async Task<string> ReplaceAsync(this Regex regex, string input, Func<Match, Task<string>> replacementFn)
```
What is the use case of this

Evengard replied:
That's used inside the humanization method, just because I needed to call async methods while regexping. It could be made private, I kept it public only because of similar methods around the class.

VelvetThePanda reviewed (changes requested):
I find this entire implementation scuffed, with the idea of this PR strange. The way half of this is done just seems purely ostentatious, given entities already contain the information necessary to do this, yet this Scribble scrabble of a route is taken to get there.

Evengard commented on DSharpPlus/Utilities.cs:
```diff
@@ -169,12 +169,154 @@ internal static IEnumerable<ulong> GetChannelMentions(DiscordMessage message)
         }
 
-        internal static IEnumerable<ulong> GetEmojis(DiscordMessage message)
+        internal static IEnumerable<(ulong id, string name, bool isAnimated)> GetGuildEmojis(DiscordMessage message)
```
I didn't thought that returning an emoji object from here would be acceptable. But if you say so...

Evengard commented on DSharpPlus/Utilities.cs:
```diff
@@ -169,12 +169,154 @@ internal static IEnumerable<ulong> GetChannelMentions(DiscordMessage message)
+                              match.Value.StartsWith("<a:"));
+        }
+
+        public static IEnumerable<DiscordEmoji> GetCommonEmojis(DiscordMessage message)
```
The same ones above were defined as IEnumerables as well, that's the only reason I did it the same. If you say so...

VelvetThePanda replied:
Then it's w/e.

Evengard commented on DSharpPlus/Utilities.cs:
```diff
@@ -169,12 +169,154 @@ internal static IEnumerable<ulong> GetChannelMentions(DiscordMessage message)
+
+        public static IEnumerable<DiscordEmoji> GetCommonEmojis(DiscordMessage message)
+        {
+            return DiscordEmoji.UnicodeEmojis.Keys
```
That's the only place I found the needed data.

Evengard commented:
I removed a bunch of smelly code (according to https://github.com/discord/discord-api-docs/blob/master/docs/Reference.md#timestamp-styles the standard emojis are supposed to be sent as Unicode characters anyway, so I removed all the standard emoji code alltogether), removed some code duplication (moving the dummy DiscordEmoji construction to DiscordEmoji) and added filtering of code tags (it is a bit tentative, as I'm not using a full-featured markdown parser but just a bunch of regexps, but still seems to work good enough for most cases).

Naamloos commented:
I feel like this is not something the library itself should handle. In what way is this addition warranted?

Evengard commented:
Tbh the only thing which really blocks me of doing it out of tree is the fact that too much stuff is marked as internal. Now, there's no way of constructing an DiscordEmoji object because, well, the constructor is internal, and it is needed for the construction of the emoji object of an emoji which the bot doesn't know about (aka an emoji from a server where the bot is not present). I think this is the only one request I really need, but I understand why it is internal. Hence, I can recreate the PR with just the `DiscordEmoji.FromGuildEmoteTagInfo` static method which would be all I need.

Naamloos commented:
A bot can not use emotes from a guild it is not in. Construction of the DiscordEmoji class is done via static methods.

Check the source code for the DiscordEmoji class. https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus/Entities/Emoji/DiscordEmoji.cs#L217-L385

VelvetThePanda commented:
> A bot can not use emotes from a guild it is not in. Construction of the DiscordEmoji class is done via static methods.
> 
> Check the source code for the DiscordEmoji class. https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus/Entities/Emoji/DiscordEmoji.cs#L217-L385

This is only partly true; components and webhook messages can use any emote, but don't quite me on what the API responds with if you try and get info about those emotes.

Neuheit commented:
It seems like this can be done without needing the internals of the lib, so I will be closing this PR.
